### ISSUE TITLE
Suggestion: Custom Type for ID and URI

### ISSUE THREAD

**Issue by winsmith** [TootSDK/TootSDK#85]
## The Problem

While beginning to work with the API, I stumbled several times over the fact that there are various ways of identifying an account:

- an `ID` such as `109385296492335026`
- a `URI` such as `@daniel@social.telemetrydeck.com`
- there's also the `acct` but I'll ignore that for now

Both `ID` and `URI` are just string types right now. This leads to confusion when plugging an `ID` into a function that requires a `URI` and vice versa. There is no check against doing such a thing, because the Swift types are the same. Only during runtime will users encounter various 404 not found errors presented to them as exceptions.

Seemingly random exceptions are not very user friendly, and having to thing about what "kind" the string I'm passing around is, is not very Swifty. It decreases the enjoyment and ease of use of TootSDK (not by a lot, but somewhat).

## The Proposal: Create a type `AccountID` and a type `AccountURI`

I think it would be very helpful if `ID` and `URI` were their own types. 

- It would prevent developers from accidentally passing one type into functions that require the other, preventing a whole class of exceptions from occurring.
- It would allow documentation for the types, where the difference and finer points can be explained, allowing for easier onboarding.
- It would reduce mental overhead when using the API, because developers would not have to keep in mind the different types of String they are passing around.
- Depending on implementation, there could even be error checking and normalization, such as normalizing `daniel@social.telemetrydeck.com` and `https://social.telemetrydeck.com/@daniel` to `@daniel@social.telemetrydeck.com`. This can make the library even more user friendly, and is also helpful to apps that use it, because error checking and normalization happens at the library level instead of the app level.

## The Implementation

Depending on how much effort we want to put into this, I can envision one of these strategies for implementation:

### Strategy 1: Type Aliases

This is an implementation strategy that isn't much work but would yield the most important of the benefits by clearly marking functions regarding the kind of identifier they require, and keeping track of the type an identifier is. It also allows for documenting the type alias, so additional explanation can be given to developers at the code and code completion level regarding what is expected of them.

It is also very easy to implement because no functions would actually change their signature, and no client code would have to change its signature either. 

The downside is that developers could still pass one type into functions that require the other without much warning. Also there is not a lot of semantics and no validity checking at all.

### Strategy 2: StringProtocol Conformance

This strategy would try as much as possible to prevent code that depends on the affected types and functions to be rewritten. To this end it would create two new types that store semantic information about their content but try to present as much as a string to the outside world as possible, using conformance to the [ExpressibleByStringLiteral](https://developer.apple.com/documentation/swift/expressiblebystringliteral) and [CustomStringConvertible](https://developer.apple.com/documentation/swift/customstringconvertible) protocols. 

This adds all the semantics and validation checking, which is poggers.

Here's a sketch of an example:

```swift 
public struct AccountURI: ExpressibleByStringLiteral, CustomStringConvertible {
    let domain: String
    let acct: String
    
    // normal initializer
    public init(acct: String, domain: String) {
        self.acct = acct
        self.domain = domain
    }
    
    // conformance to ExpressibleByStringLiteral
    public init(stringLiteral value: String) {
        // actual implementation here could 
        // check for various representations 
        // and extract the correct data, or throw
        let parts = value.split(separator: "@")
        self.acct = String(parts.first)
        self.domain = String(parts.last)
    }
    
    // conformance to CustomStringConvertible
    public var description: String {
        return "@\(acct)@\(domain)"
    }
}
let myAccount: AccountURI = "@daniel@social.telemetrydeck.com"
print(myAccount) // "@daniel@social.telemetrydeck.com"
````

The main downside is that even with `ExpressibleByStringLiteral` and `CustomStringConvertible`, this strategy would still break people's code somewhat, and it is trying to do an awful lot of magic which might be confusing in the long run.

### Strategy 3: Just break it all and go for proper structs

This strategy is the most disruptive. It would create new structs for `URI` and `ID`, same as Strategy 2, but not go for any conformance with pre-existing APIs. Code will just break and people would have to deal with it.

**Comment by kkostov** [comment]
@winsmith this is an excellent write-up.

I agree that TootSDK should solve this for devs, especially when deciding which “identifier” to provide is usually unclear from the docs and also more confusing when you consider that different server types have different expectations. For example (plot twist), Pleroma `/api/v1/accounts/:id` and `/api/v1/accounts/:id/statuses` accept both ID and nickname!

From the listed options, I like options 2 (`ExpressibleByStringLiteral`, `CustomStringConvertible`) and 3 (proper struct) the most.

Pre-v1, we are not really looking for total API stability, especially when the gains are significant, which I think they are in this case. What I like is that in both options is that we get a proper struct, so we can implement flavour checks inside it, so we can handle any server-specific implementations.

For example, `func getAccount(by id: AccountID)` and an overload `func getAccount(by nickname: AccountNickname)` where in the latter, we can detect whether the flavour is Pleroma or not and resolve the nickname to an ID.

Another aspect I really like is that we can introduce a more refined naming for these parameters that's unified across server types (so the entire TootSDK will refer to `AccountID` and `AccountNickname`, `AccountURI` etc. We did a similar "unification" with the term `Post` for example (so you don't see a mix of post/toot/status).

**Comment by davidgarywood** [comment]
Yup, what I _really_ want is for the app not to have to care too much about this. I think option 2 with a few bit stops smooth things out, probably gets us there. Like @kkostov I don't care at this stage about breaking API stability, if it means we get something really good out of it!

**Comment by winsmith** [comment]
Aight, I'll draft up a more refined suggestion! Thanks for your feedback!
